Print message 2

casperdcl opened the pull request:
Rebased and fixed as per #138 , #148

casperdcl commented:
@lrq3000 @graingert @cscanlin are you happy with this? Haven't done much testing myself since I don't need this feature.

graingert commented on README.rst:
```diff
@@ -436,6 +451,28 @@ For manual control over positioning (e.g. for multi-threaded use),
+    for i in bar:
+        # Print using tqdm class method .write()
+        tqdm.write("Done task %i" % i)
+        # Can also use instance method bar.write()
```
that's not an instance method. It's still a class method.

casperdcl replied:
Yes, well, slightly more accurately: a class method called via an instance. Will reword that in addition to anything else this weekend.

lrq3000 replied:
Checking this evening.

2016-04-22 16:31 GMT+02:00 Casper da Costa-Luis <notifications@github.com>:

> In README.rst <https://github.com/tqdm/tqdm/pull/150#discussion_r60747984>
> :
>
> > +Since ``tqdm`` uses a simple printing mechanism to display progress bars,
> > +you should not write any message in the terminal using ``print()``.
> > +
> > +To write messages in the terminal without any collision with ``tqdm`` bar
> > +display, a ``.write()`` method is provided:
> > +
> > +.. code:: python
> > +
> > +    from tqdm import tqdm, trange
> > +    from time import sleep
> > +
> > +    bar = trange(10)
> > +    for i in bar:
> > +        # Print using tqdm class method .write()
> > +        tqdm.write("Done task %i" % i)
> > +        # Can also use instance method bar.write()
>
> Yes, well, slightly more accurately: a class method called via an
> instance. Will reword that in addition to anything else this weekend.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tqdm/tqdm/pull/150/files/8e1ce8bf117b8b2b7078390aee7488fcd34d053a#r60747984>
>

lrq3000 replied:
@graingert You're right but I wanted to make things simple to understand for the reader, and also to get a shorter comment (flake8!). But yeah, the goal was just to point that .write() could be used both from `tqdm` class or from an instance of `tqdm`.

graingert replied:
But that's obvious if it's a class method
On 22 Apr 2016 7:08 pm, "Stephen L." <notifications@github.com> wrote:

> In README.rst <https://github.com/tqdm/tqdm/pull/150#discussion_r60779668>
> :
>
> > +Since ``tqdm`` uses a simple printing mechanism to display progress bars,
> > +you should not write any message in the terminal using ``print()``.
> > +
> > +To write messages in the terminal without any collision with ``tqdm`` bar
> > +display, a ``.write()`` method is provided:
> > +
> > +.. code:: python
> > +
> > +    from tqdm import tqdm, trange
> > +    from time import sleep
> > +
> > +    bar = trange(10)
> > +    for i in bar:
> > +        # Print using tqdm class method .write()
> > +        tqdm.write("Done task %i" % i)
> > +        # Can also use instance method bar.write()
>
> @graingert <https://github.com/graingert> You're right but I wanted to
> make things simple to understand for the reader, and also to get a shorter
> comment (flake8!). But yeah, the goal was just to point that .write() could
> be used both from tqdm class or from an instance of tqdm.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tqdm/tqdm/pull/150/files/8e1ce8bf117b8b2b7078390aee7488fcd34d053a#r60779668>
>

lrq3000 replied:
In Python, maybe. But it was not obvious to me at first. I explicitly wanted both behaviors to be possible, Python makes that easy, but still I prefer that behavior to be explicit rather than implicit.

lrq3000 commented on tqdm/_tqdm.py:
```diff
@@ -561,7 +581,7 @@ def __iter__(self):
                         elapsed = cur_t - start_t
                         # EMA (not just overall average)
-                        if smoothing:  # and delta_it
+                        if smoothing and delta_t:
```
Why are you restoring `delta_t`? Is it safe? I remember we had some issues in the past with `delta_it` (but don't remember the details - should do a git blame?)...

lrq3000 commented:
Except the "Revert minor optimization" commit which I'm not too sure about (see my comment above), this PR seems alright. I didn't run it but I checked the sourcecode with my original PR and it's the same (and I extensively tested my original PR).

This should fix #113 and #116 alright.

lrq3000 commented:
BTW, I take the opportunity of this PR to point to a clone of `tqdm` in ruby!

https://github.com/powerpak/tqdm-ruby

Do you think we should point to them in the readme? More `tqdm`, happier the world will be! ;p

casperdcl commented:
@lrq3000 ok, created a separate PR and rebased. (also rebased the two PRs which depend on this one. they depend on this one purely to push us to tidy this and merge it in.)

I know about the ruby port... once it's tidy enough @powerpak (particularly regression testing) we could host it on this org.

casperdcl commented:
BTW: I just changed the repo codecov options so that codecov/patch is required, while codecov/project is not (rather than the other way around).

lrq3000 commented:
Great Casper, thank you very much! I am also waiting for this PR to be
merged to finish the IPython module. Didn't think this PR would be so
useful when I first made it!

Yes I agree about the ruby port, but I really hope to see its progress :)

About codecov, I don't know the difference between patch and project lol,
what does it mean in practice?

2016-04-23 0:29 GMT+02:00 Casper da Costa-Luis <notifications@github.com>:

> BTW: I just changed the repo codecov options so that codecov/patch is
> required, while codecov/project is not (rather than the other way around).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tqdm/tqdm/pull/150#issuecomment-213611897>
>

casperdcl commented:
patch checks our requirement for hitting 80% overall coverage (can't remember if you set that limit... I recall there were reasons along the lines of it being high but not so high that medium sized new beta features would fail). project is just silly if we have approx 100% coverage. it checks that overall coverage does not decrease.

casperdcl commented:
Not quite there yet:

``` python
# coding: utf-8
"""
output:
i:   0%|                                                  | 0/8 [00:00<?, ?it/s]
Done task 0
i:  50%|█████████████████████                     | 4/8 [00:05<00:05,  1.30s/it]
Done task 4███████████████████████████████████████| 8/8 [00:01<00:00,  6.14it/s]
i: 100%|██████████████████████████████████████████| 8/8 [00:10<00:00,  1.30s/it]
j: 100%|██████████████████████████████████████████| 8/8 [00:01<00:00,  6.13it/s]
k: 100%|████████████████████████████████████████| 16/16 [00:00<00:00, 98.46it/s]
"""

from __future__ import print_function
from tqdm import tqdm, trange, tgrange
from time import sleep


for i in trange(8, leave=True, desc='i'):
    if not (i % 4):
        tqdm.write("Done task %i" % i)
    for j in trange(8, leave=True, desc='j'):
        for k in trange(16, leave=True, desc='k'):
            sleep(0.01)
print('\n')
```

lrq3000 commented:
Ok I'll look into that this evening, there must be something that broke the
feature with the latest tqdm changes

lrq3000 commented:
Oh wow, indeed the output is totally messed up, even when I'm using my test scripts (that worked perfectly fine before). I'll see what I can do

lrq3000 commented on tqdm/_tqdm.py:
```diff
@@ -282,6 +285,23 @@ def _decr_instances(cls, instance):
+        """
+        # Clear all bars
+        for inst in cls._instances:
+            if inst.fp == file:
```
This is what caused the bug. In fact, the `.write()` method wasn't working at all, simply because the default `file` for `.write()` was set to `sys.stdout` while `tqdm` progress bar is set to `file=sys.stderr` by default.
In the past it worked, because `.write()` would work on all bars whatever the file. I figured  `sys.stdout` was the most reasonable choice for most people, they don't want to print to `sys.stderr`, but well, I reverted to agree with `tqdm`.

lrq3000 replied:
Well, no, I don't agree, in fact there is a collusion between `sys.stdout` and `sys.stderr` (they both print in the terminal as the same kind of content), so I will change that so that `.write()` will clean both `sys.stdout` and `sys.stderr`. I guess this is the safest choice for most people, and the less confusing.

lrq3000 commented:
Ok done, let me think what you think of my choice (ie, clearing only bars in output file OR if output file is either stdout or stderr, then clear both stdout and stderr).
I just need to update the unit test to get 100% coverage and it's done.

casperdcl commented:
Thinking about CLI usage (where stderr is progress output on console and
stdout is piped to another program/logfile) might be helpful here. If
someone wants to print a message in a tqdm-wrapped loop It's probably to do
with progress, which means it's probably good to put it in the same file
that `tqdm` itself is using for its messages - i.e. get the file for the
current instance (defaulting to stderr)

On 23 April 2016 at 22:57, Stephen L. <notifications@github.com> wrote:

> Ok done, let me think what you think of my choice (ie, clearing only bars
> in output file OR if output file is either stdout or stderr, then clear
> both stdout and stderr).
> I just need to update the unit test to get 100% coverage and it's done.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/tqdm/tqdm/pull/150#issuecomment-213840890>
>

lrq3000 commented:
Thank you for your input @casperdcl, however I disagree, because per #113, the original discussion, the goal was to provide a way for user to print messages in the terminal without them being mixed up with tqdm output. The implicit need was to be able to print to sys.stdout without it being mixed up with tqdm which prints in sys.stderr (but the console mix up sys.stdout and sys.stderr), hence my proposition.

I have reverted the change for the moment, it can be merged, but I expect this issue to pop up in the future, so we should keep this in mind...

casperdcl commented:
Hmm @lrq3000 I don't really have a strong opinion on this as I've never needed this feature. Happy to go with whatever you think best. I agree that in practice it makes sense to clear all bars with `file in sys.(stdout|stderr)` since if either is redirected to a file people are unlikely to care about the extra control characters.

lrq3000 commented:
I don't need this option either, but I want it to be done for good, that was pretty complicated to come up with a solution to this problem :-p

About the extra control characters, no there won't be, because if for example the user is redirecting stdout to a file, and prints stderr to console, then only the messages will be written in stdout, no the bars refresh, so no extra control characters even in this case. In fact, the only drawback to `file in sys.(stdout|stderr)` is that we may be refreshing bars in stderr for nothing if user is redirecting stdout to a file, but that's all, only a few cycles and IO lost at worst.

Ok then I revert again back the commit and add a unit test!

lrq3000 commented:
Reverted back and added more unit testing. If Travis and Codecov pass, this PR is done!

lrq3000 commented:
Rebased to fix conflicts. Please review my changes. Can be merged if Travis and Codecov pass.

lrq3000 commented:
Wow fast @casperdcl XD

casperdcl commented:
well I squashed a few commits and merged without conflicts (clearly fast-forward) so no point in waiting for tests to finish... naturally the perf test failed and ~~is currently resuming~~ has just passed

lrq3000 commented:
lol as always when we merge a new PR...

2016-04-25 22:47 GMT+02:00 Casper da Costa-Luis <notifications@github.com>:

> well I squashed a few commits and merged without conflicts (clearly
> fast-forward) so no point in waiting for tests to finish... naturally the
> perf test failed and is currently resuming has just passed
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tqdm/tqdm/pull/150#issuecomment-214516531>
>
